### ISSUE TITLE
crossnobis reliability

### ISSUE THREAD

**Issue by psychNerdJae** [rsagroup/rsatoolbox#239]
When computing crossnobis distances, the resulting neural RDM contains a diagonal that consists entirely of zeroes. The behavior I might have instead expected is for the diagonal to instead contain information about the "reliability" of each pattern across runs/crossvalidation folds – e.g., so that I can verify whether the pattern distance for stimulus 1 is more similar than the pattern distance for stimulus 1 vs stimulus 2.

It seems the diagonal is "blanked out" at [this step](https://github.com/rsagroup/rsatoolbox/blob/08930664bf6d1724575727f088278ce2bb4ab6dc/rsatoolbox/rdm/calc.py#L442), which uses [some variant of this algorithm](https://math.stackexchange.com/questions/2240429/pairwise-distance-matrix) to vectorize the computation of the pairwise distance matrix. And even if this were not the case, the subsequent extraction of the upper triangle would remove the diagonal anyways. I understand that the vectorized approach has [computational benefits](https://github.com/rsagroup/rsatoolbox/issues/181), but I'm also curious whether it comes at the detriment of losing information about reliability?

**Comment by JasperVanDenBosch** [comment]
Hi @psychNerdJae thanks for your question. I'll give a short answer with my understanding but other may want to weigh in . This is a good question and has been debated for some time in various places, see for example also [this paper](https://doi.org/10.1016/j.neuroimage.2016.12.079).

You are right ofcourse that there are benefits to having access to the values of the diagonal for various purposes. Therefore perhaps we should consider returning these values, perhaps as a separate variable. However I think that for most RSA purposes the definition of an RDM is explicitly such that the diagonal is empty, not just for performance sake but more for convenience in successive steps of the analysis, and for clarity.

**Comment by nkriegeskorte** [comment]
there are two fundamentally different ways to use data splitting to remove the bias of dissimilarity estimates.

one approach (which you are thinking of) is the split-data-RDM approach, where one data set is used for the rows and the other for the columns of the RDM and a *biased* dissimilarity estimator is then used to compute distances between patterns. for each dissimilarity, one condition's pattern is from data set A and the other condition's pattern from data set B. each entry of the diagonal has the dissimilarity between two patterns corresponding to the same condition in the two data sets, reflecting the pattern-estimate reliability. in the past, we sometimes substracted the average of the diagonal of the split-data-RDM from each off-diagonal entry to remove the bias. however, this approach has drawbacks as described in nili et al. 2020.

the crossnobis estimate is not an example of the split-data RDM approach, because it does not use one data set for the rows and the other for the columns. it uses the difference vector between the two conditions in data set A and the difference vector between the two conditions in data set B, and is by definition exactly zero on the diagonal (see eq 3 in kriegeskorte & diedrichsen 2019). this is a better approach to remove the bias.

**Comment by psychNerdJae** [comment]
Hi @ilogue, thanks for pointing me towards that paper – the argument spelled out there is pretty much my exact rationale for wanting to check the diagonal for reliability. I appreciate your point that for most RSA applications, the upper/lower triangle is the main quantity of interest. Certainly in my case, I'd only be interested in the diagonal to assess reliability – and then I'd be returning right back to the upper triangles when comparing the neural and model RDMs.

Hi @nkriegeskorte, thanks for the thoughtful clarification, and for the pointer to the annual reviews paper. After reviewing the crossnobis equations, it's clear that I was conceptually confused on this point.

I have a potentially naive follow-up question about reliability, where I'd benefit from your expert advice. To our surprise, after using crossnobis to estimate neural RDMs in hippocampus, we're finding that many of our stimuli seem to be represented very similarly to each other (most values hovering near zero, and also many negative values). One interpretation of this is that many of those true distances are near-zero, and we're simply seeing that the crossnobis is returning noisy estimates of the true distance. We reasoned that there might be another possibility, which is that our measurements could be noisy enough that the same stimulus seems to evoke very different patterns across runs. If the reliability of our pattern measurements is not especially good, then it might be the case that interpreting off-diagonal dissimilarity is a fool's errand anyways. The question is this: has there been any theoretical work examining what kinds of neural RDMs result from measuring patterns that are highly noisy across runs, and if so, is it possible that this might produce the kinds of RDMs we're finding?

**Comment by nkriegeskorte** [comment]
you're on the right track. if patterns are unreliable across runs, then condition pairs will not be linearly decodable and crossnobis (or LD-t) dissimilarities will be symmetrically distributed about 0. if you compute LD-t values (kriegeskorte et al. 2007, nili et al. 2014) you can test easily for significantly positive pattern dissimilarities.

**Comment by psychNerdJae** [comment]
This is very helpful, thanks – so my understanding is that we could simply extract the upper triangle of the neural RDM and do a t-test on that vector. If the average value is significantly above 0, then this suggests that the pairwise distances are decodable, otherwise we need to take a step back and evaluate why we're unable to measure reliable patterns in a given region. I'll go ahead and close this issue with the assumption that my understanding is correct, but please feel free to open this back up if I've misunderstood something.

I don't know if there's any maintainer who works on the readthedocs documentation, but if they're interested in user feedback (from the point of view of someone who's just starting to work w/ neural RSA) – while the page on [computing distances](https://rsatoolbox.readthedocs.io/en/latest/distances.html#crossnobis-dissimilarity) contains all of the information discussed in this conversation, it's stated in a way that's a little bit hard to follow.

For example:
> If the true distance is zero (i.e. if two patterns only differ by noise), the average estimated distance will be zero. If there is not information in a set of activity patterns, then half the distance estimates will be positive, and half the estimates will be negative.

At face value, these seem like mutually contradictory statements, as it suggests that an estimated distance of zero reflects both a true distance of zero, but also a lack of information. I now understand "set of activity patterns" to mean that the *entire* neural RDM should have a positive average value (where positive values are a proxy for decodability), with the occasional set of near-zero activity patterns when two stimuli are represented very similarly. I wonder if the documentation could be revised to make this clearer? This point could potentially be expanded upon in the first bullet point starting with "You can perform a t-test of the crossnobis estimates against zero, exactly like you would test the classification accuracy of a decoder against chance performance."

Thanks all for the quick and thoughtful replies, much appreciated!

**Comment by nkriegeskorte** [comment]
the t-test i refer to is that of the LD-t, described in kriegeskorte et al. 2007 and nili et al. 2014. (can't t-test the vector of dissimilarities because of dependency.)

"no information in the activity patterns" means no mutual information between stimuli and response patterns. this is the case when the patterns are drawn from the same distribution, no matter what the stimulus was. in that case unbiased dissimilarity estimates will be zero on average.
